File: sdks/php/src/Pkce.php
<?php

declare(strict_types=1);

namespace Dalang\Sso;

/** A PKCE verifier/challenge pair (RFC 7636, S256). */
final class Pkce
{
    public function __construct(
        public readonly string $verifier,
        public readonly string $challenge,
    ) {
    }

    /** Generate a random verifier and derive its S256 challenge. */
    public static function generate(): self
    {
        $verifier = self::b64url(random_bytes(32));
        // challenge = base64url(sha256(verifier)) with no padding.
        $challenge = self::b64url(hash('sha256', $verifier, true));

        return new self($verifier, $challenge);
    }

    /** base64url-encode without padding. */
    private static function b64url(string $raw): string
    {
        return rtrim(strtr(base64_encode($raw), '+/', '-_'), '=');
    }
}

File: sdks/go/pkce.go
package dalangsso

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
)

// Pkce is a PKCE verifier/challenge pair (RFC 7636, S256).
type Pkce struct {
	Verifier  string
	Challenge string
}

// GeneratePkce creates a random verifier and derives its S256 challenge:
// challenge = base64url(sha256(verifier)) with no padding.
func GeneratePkce() (*Pkce, error) {
	buf := make([]byte, 32)
	if _, err := rand.Read(buf); err != nil {
		return nil, err
	}
	verifier := base64.RawURLEncoding.EncodeToString(buf)
	sum := sha256.Sum256([]byte(verifier))
	challenge := base64.RawURLEncoding.EncodeToString(sum[:])
	return &Pkce{Verifier: verifier, Challenge: challenge}, nil
}
